cmd/runner/main.go

```diff
@@ -58,13 +73,33 @@ func main() {
 
 	done := make(chan string)
 
+	suitesReporter := runner.NewTestSuitesReporter(outputSuiteName)
+
+	suitesReporter.SetStartTime(time.Now())
+
 	for qName, configs := range configQueueMap {
-		reporter := runner.NewTestSuiteReporter(qName, logPrefixFmt)
-		go r.Run(configs, reporter, c[qName], done)
+		suiteReporter := suitesReporter.NewTestSuiteReporter(qName, logPrefixFmt)
+		go r.Run(configs, suiteReporter, c[qName], done)
 	}
 
 	for range configQueueMap {
 		qName := <-done
 		log.Printf("Done running tests for queue %q", qName)
 	}
+
+	suitesReporter.SetEndTime(time.Now())
+
+	if len(o) > 0 {
```

Comment:
There are multiple options here:

1. Should we always generate an xml report, and print it to stdout if no output file is specified?
2. Should we always generate an xml report, and print it to a file only if it is specified?
3. Should we generate an xml report only if an output file is specified?

Right now we are following option 2. It seems wasteful to create a report that is not used.

What we could do instead is create the "suites" report in the main file, and pass it as an argument when creating the suitesReporter. This would trigger the creation of "suite" and "case" reports at the lower levels.

If we do that, we will need to keep a testCaseCount in the suiteReporter, since the suite report itself may be nil and not keep a list of case reports.

We might get maximum flexibility by having an argument to pass in a suiteReport when creating a suiteReporter as well. That way we could explicitly create a report only for testSuites for which a report is requested.

Reply:
I do not think we should generate an XML report unless it is specified. I believe the driver does not at present. I agree that (2) has unnecessary data structure creation. That is why I'm not sure about the coupling of the logger and the junit report. To me, (3) seems most appropriate.

Reply:
Choosing (3), see below for implementation discussion and details.